java/c/src/main/java/org/apache/arrow/c/ArrayImporter.java

```diff
@@ -138,16 +139,143 @@ private List<ArrowBuf> importBuffers(ArrowArray.Snapshot snapshot) {
         buffersCount, vector.getField().getType().getTypeID(), buffers.length);
 
     List<ArrowBuf> result = new ArrayList<>(buffersCount);
+    int bufferIndex = 0;
     for (long bufferPtr : buffers) {
       ArrowBuf buffer = null;
       if (bufferPtr != NULL) {
         // See ARROW-17720: [Java] C data interface: Add API to compute imported buffer size
-        int capacity = Integer.MAX_VALUE;
-        buffer = new ArrowBuf(referenceManager, null, capacity, bufferPtr);
+        int capacity = vector.getField().getType().accept(new IntegerArrowTypeVisitor(fieldNode, bufferIndex++));
+        AllocationManager allocationManager = new CDataAllocationManager(allocator, underlyingAllocation, bufferPtr, capacity);
+        allocator.forceAllocate(capacity);
+        BufferLedger ledger = new BufferLedger(allocator, allocationManager);
+        buffer = new ArrowBuf(ledger, null, capacity, bufferPtr);
         buffer.writerIndex(capacity);
       }
       result.add(buffer);
     }
     return result;
   }
+
+  private static class IntegerArrowTypeVisitor implements ArrowType.ArrowTypeVisitor<Integer> {
```

Comment:
We can compute buffer lengths based on the type. Obviously, this needs to be factored out and tested and made robust and all that.

Reply:
Needs some refactoring to properly handle varwidth types